Declare browser autonomy as an explicit goal (#9)
hober/web-extensions: charter.md

## Goals

### Autonomy

Individual browser vendors are free to implement browser extensions in whatever
way they feel is best for their end-users and extension developers. Our
processes embrace this and seek to provide mechanisms for specifying
inconsistencies and working towards eventual unification where appropriate. We
also recognize that not all APIs or capabilities need to be specified.

The need for autonomy must be balanced against the need to provide developers
with a consistent platform. To that end, we seek to specify a common platform
that includes the base extension model, permissions model, and a common core of

> APIs for web extensions that all browsers can build upon.

Reviewer: I think "that all browsers can build upon" needs to be a little stronger, especially to balance out the first sentence of  "Individual browser vendors are free to implement browser extensions in  whatever way they feel is best".

Author: Please consider everything I've written so far to be draft language. I'm not beholden to any particular wording I've used here. 

One of the issues I was grappling with as I was drafting this is that it's not clear to me how much a specification or standard can demand of a browser vendor. This may be ignorance on my part, but I'm not aware of any formal commitment that browsers make to implementing any given spec or standard. I chose the phrase "that all browsers can build upon" because of a mix of that concern and the previously discussed desire for us to start with a small core of APIs, systems, and concepts that we specify as web extensions.

I still don't have a good idea on how we could better frame this, but I'm totally open to suggestion. Do you have any opinions on how we could strengthen this wording or rewrite it?

Reviewer 2: Maybe start with something like "We recognize that browser vendors are in need of a mechanism to experiment with new features and provide functionality that is specific to the respective browser. Our processes embrace this...."

At least in the specification phase I don't feel there would be any formal commitment to make changes in accordance to what is specified. This is slightly more the case if we eventually standardize. But even then, a browser vendor may choose to disregard the standard. It would mainly be a matter of public perception as to what that means in the long run.

Reviewer 3: I'm questioning the need to specifically call out "Browser vendors can do what they want".  I'm not familiar with a similar "Autonomy" principle for other web-related standards, and yet there's a perception that Google (Chrome) already behaves in a similar way (for example, see [quotes](https://twitter.com/anssik/status/1313862250868940801) from the recent "Investigation of Competition in Digital Markets" by a Committee of the U.S Congress).

> 

Reviewer 2: I think what I would like to see in this section is a commitment that browser vendors are open to making changes to already shipped features if the spec/standard we agree on would require it. While this may have an additional impact on the vendor to migrate developers, not doing so puts pressure on the other vendors to accept status quo because there is already a shipped implementation. I'm hoping that through the spec and future standard, we can bring the best possible extension model and APIs to developers together.

> 

Reviewer 2: I think what I would like to see in this section is a commitment that browser vendors are open to making changes to already shipped features if the spec/standard we agree on would require it. While this may have an additional impact on the vendor to migrate developers, not doing so puts pressure on the other vendors to accept status quo because there is already a shipped implementation. I'm hoping that through the spec and future standard, we can bring the best possible extension model and APIs to developers together.

Our specifications should not make reference to or rely on specific
browser engine implementation details.

### Autonomy

We recognize that browser vendors need to provide functionality that is specific
to their browser and also need the ability to experiment with new features. Our
process embraces this and seeks to provide mechanisms for specifying
inconsistencies and working towards eventual unification where appropriate.

> Given this, we expect browser vendors to support APIs and capabilities that we
> do not specify.

Author: ```suggestion
Given this, we expect browsers will support APIs and capabilities that we do not
specify and may iterate on specified APIs.
```

While reviewing this PR earlier today, one of our extensions engineers called out that the current framing does not address potential divergences between specified behavior and a browser's implementation. This suggestion attempts to address that gap.

Author: In today's sync the concerns were raised that this suggestion stands in direct contrast to the goal of providing developers with a consistent, stable, cross-browser platform. Given these concerns, I'm dropping this suggestion.

> Given this, we expect browser vendors to support APIs and capabilities that we
> do not specify.

Reviewer 4: ```suggestion
Given this, we expect browser vendors to offer APIs and capabilities beyond what's specified.
```

(changed wording to capture the original meaning plus the desire to emphasize that vendors may iterate on existing APIs.)
